src/ubx.cpp

```diff
@@ -308,63 +310,88 @@ GPSDriverUBX::configure(unsigned &baudrate, OutputMode output_mode)
-	_buf.payload_tx_cfg_tmode3.flags        = 0; /* disable time mode */
+    // disable RTCM output
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1005, 0);
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1077, 0);
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1087, 0);
 
-	if (!sendMessage(UBX_MSG_CFG_TMODE3, (uint8_t *)&_buf, sizeof(_buf.payload_tx_cfg_tmode3))) {
-		return -1;
-	}
+    memset(&_buf.payload_tx_cfg_tmode3, 0, sizeof(_buf.payload_tx_cfg_tmode3));
+    _buf.payload_tx_cfg_tmode3.flags = 0; /* disable time mode */
 
-	if (waitForAck(UBX_MSG_CFG_TMODE3, UBX_CONFIG_TIMEOUT, true) < 0) {
-		return -1;
-	}
+    sendMessage(UBX_MSG_CFG_TMODE3, (uint8_t *)&_buf, sizeof(_buf.payload_tx_cfg_tmode3));
+    waitForAck(UBX_MSG_CFG_TMODE3, UBX_CONFIG_TIMEOUT, true);
 
+    return 0;
+}
+
+int GPSDriverUBX::restartSurveyIn()
+{
 	UBX_DEBUG("Starting Survey-in");
 
 	memset(&_buf.payload_tx_cfg_tmode3, 0, sizeof(_buf.payload_tx_cfg_tmode3));
 	_buf.payload_tx_cfg_tmode3.flags        = UBX_TX_CFG_TMODE3_FLAGS;
 	_buf.payload_tx_cfg_tmode3.svinMinDur   = _survey_in_min_dur;
 	_buf.payload_tx_cfg_tmode3.svinAccLimit = _survey_in_acc_limit;
 
-	if (!sendMessage(UBX_MSG_CFG_TMODE3, (uint8_t *)&_buf, sizeof(_buf.payload_tx_cfg_tmode3))) {
-		return -1;
-	}
+    sendMessage(UBX_MSG_CFG_TMODE3, (uint8_t *)&_buf, sizeof(_buf.payload_tx_cfg_tmode3));
+    waitForAck(UBX_MSG_CFG_TMODE3, UBX_CONFIG_TIMEOUT, true);
 
-	if (waitForAck(UBX_MSG_CFG_TMODE3, UBX_CONFIG_TIMEOUT, true) < 0) {
-		return -1;
-	}
+    decodeInit();
 
-	/* enable status output of survey-in */
-	if (!configureMessageRateAndAck(UBX_MSG_NAV_SVIN, 5, true)) {
-		return -1;
-	}
+    // disable RTCM output
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1005, 0);
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1077, 0);
+    configureMessageRateAndAck(UBX_MSG_RTCM3_1087, 0);
 
 	return 0;
 }
 
+int GPSDriverUBX::enableFixedLLA()
+{
+    UBX_DEBUG("Setting fixed LLA");
+
+    memset(&_buf.payload_tx_cfg_tmode3, 0, sizeof(_buf.payload_tx_cfg_tmode3));
+    _buf.payload_tx_cfg_tmode3.flags        = UBX_TX_CFG_TMODE3_FIXEDLLA;
+    _buf.payload_tx_cfg_tmode3.ecefXOrLat   = _fixed_survey_latitude;
+    _buf.payload_tx_cfg_tmode3.ecefYOrLon   = _fixed_survey_longitude;
+    _buf.payload_tx_cfg_tmode3.ecefZOrAlt   = _fixed_survey_altitude;
+    _buf.payload_tx_cfg_tmode3.ecefXOrLatHP = _fixed_survey_latitudeHP;
+    _buf.payload_tx_cfg_tmode3.ecefYOrLonHP = _fixed_survey_longitudeHP;
+    _buf.payload_tx_cfg_tmode3.ecefZOrAltHP = _fixed_survey_altitudeHP;
+    _buf.payload_tx_cfg_tmode3.fixedPosAcc  = 1;
+    _buf.payload_tx_cfg_tmode3.svinMinDur   = _survey_in_min_dur;
+    _buf.payload_tx_cfg_tmode3.svinAccLimit = _survey_in_acc_limit;
+
+    sendMessage(UBX_MSG_CFG_TMODE3, (uint8_t *)&_buf, sizeof(_buf.payload_tx_cfg_tmode3));
+    waitForAck(UBX_MSG_CFG_TMODE3, UBX_CONFIG_TIMEOUT, true);
+
+    decodeInit();
+
+    // enable RTCM output
```

Comment:
Enable/disable is repeated a couple of places - maybe a good DRY candidate.

Reply:
I thought about doing that, but I decided against it for mainly one reason: this piece of code is safety-critical, and if RTCM streaming doesn't absolutely get shut off if the driver isn't in fixed-position mode, our RTK solution goes out the window and presently the GCS or Vehicle doesn't have a way to know what's wrong. I wanted to make it impossible for that to happen.